backend/tournesol/views/inconsistencies.py

```diff
@@ -8,12 +8,188 @@
+
+        all_connections: all the edges (also including comparisons with a score of 0 as edges)
+        """
+        self.all_entities = set()
+        self.outgoing_connections = {}
+        self.incoming_connections = {}
+
+        # all the edges (also including comparisons with a score of 0 as edges)
+        self.all_connections = {}
+
+        criterion_comparisons = list(
+            filter(lambda d: d["criteria"] == self.criterion, comparisons)
+        )
+
+        for comparison in criterion_comparisons:
+            entity_1 = comparison["comparison__entity_1__uid"]
+            entity_2 = comparison["comparison__entity_2__uid"]
+
+            if entity_1 not in self.all_entities:
+                self.all_entities.add(entity_1)
+                self.incoming_connections[entity_1] = set()
+                self.outgoing_connections[entity_1] = set()
+                self.all_connections[entity_1] = set()
+
+            if entity_2 not in self.all_entities:
+                self.all_entities.add(entity_2)
+                self.incoming_connections[entity_2] = set()
+                self.outgoing_connections[entity_2] = set()
+                self.all_connections[entity_2] = set()
+
+            if comparison["score"] > 0:
+                # Entity_2 is preferred
+                self.incoming_connections[entity_2].add(entity_1)
+                self.outgoing_connections[entity_1].add(entity_2)
+
+            elif comparison["score"] < 0:
+                # Entity_1 is preferred
+                self.incoming_connections[entity_1].add(entity_2)
+                self.outgoing_connections[entity_2].add(entity_1)
+            else:
+                continue
+
+            self.all_connections[entity_1].add(entity_2)
+            self.all_connections[entity_2].add(entity_1)
+
+    def _count_cycles_and_comparison_trios(self):
+        """
+        Count the cycles of length 3 and the comparison trios.
+        Add the cycles found to the cycles list.
+
+        The cycles returned are in the order 1, 2, 3 with 1 > 2 > 3 > 1.
+        """
+        cycles_count = 0
+        comparison_trios_count = 0
+        for entity_1 in self.all_entities:
+            # Search the cycles
+            for entity_2 in self.outgoing_connections[entity_1].copy():
+                for entity_3 in self.incoming_connections[entity_1] & \
+                                self.outgoing_connections[entity_2]:
+                    cycles_count += 1
+                    self.cycles.append({
+                        "criterion": self.criterion,
+                        "entity_1_uid": entity_1,
+                        "entity_2_uid": entity_2,
+                        "entity_3_uid": entity_3,
+                    })
+
+                # Remove the edge, so that these cycles won't be counted again
+                self.outgoing_connections[entity_1].remove(entity_2)
+                self.incoming_connections[entity_2].remove(entity_1)
```

Comment:
Mmmm, I see. Having these operations here makes it much harder for me to keep track of what's hapening during the multiple loops.

I would find something that does not modify the list of connections but keeps track of a `visited` object simpler to reason about.

What do you think?

Reply:
There were 2 ways I think to do this:
1 - using simple for loops and if (the suggestion of the previous comment), which is better for readability 
2 - using the "&" operator on sets, which is better for performances

I don't know if I can keep using the 2nd method without progressively "removing the edges" where all the cycles were counted, even if it's not great for readability. So I believe it is a choice between these 2 techniques.

Reply:
I agree that using immutable connections here would help to reason about the algorithm, especially for future readers and maintainers.

If "removing the edges" is only used to avoid counting the same cycle multiple times, then an alternative solution would be to consider only the cycles where `entity_1` has the smallest uid among the 3 (i.e adding conditions `entity_1 < entity_2` and `entity_1 < entity_3` when iterating on the connections).

(This also applies to "comparison trios").

Reply:
For comparisons trios, I applied a different trick. It turns out that, without removing the edges, each cycle is counted 3 times, and each comparisons trio is counted 6 times. For the comparisons trios I only need a count, so I can just divide it by 6, which is more efficient than incrementing the count only when uid1 < uid2 < uid3.